packages/web/src/components/Form.js

```diff
@@ -0,0 +1,191 @@
+// Module imports
+import {
+	createContext,
+	useCallback,
+	useContext,
+	useEffect,
+	useReducer,
+} from 'react'
+import PropTypes from 'prop-types'
+
+
+
+
+
+// Local imports
+import {
+	auth,
+	firestore,
+} from 'helpers/firebase'
+
+
+
+
+
+// Constants
+const INITIAL_STATE = {
+	errors: {},
+	isValid: false,
+	isTouched: false,
+	touched: {},
+	validity: {},
+	values: {},
+}
+
+
+
+
+
+function reducer(state, action) {
+	const {
+		payload,
+		type,
+	} = action
+	const newState = {
+		...INITIAL_STATE,
+		...state,
+	}
+
+	switch (type) {
+		case 'validity changed':
+			newState.validity = {
+				...newState.validity,
+				[payload.fieldName]: !payload.errors?.length,
+			}
+			newState.errors = {
+				...newState.errors,
+				[payload.fieldName]: payload.errors,
+			}
+			newState.isValid = !Object
+				.values(newState.validity)
+				.some(isValid => !isValid)
+			break
+
+		case 'value changed':
+			newState.values = {
+				...newState.values,
+				[payload.fieldName]: payload.value,
+			}
+			newState.touched = {
+				...newState.touched,
+				[payload.fieldName]: newState.initialValues[payload.fieldName] !== payload.value,
+			}
+			newState.isTouched = Object
+				.values(newState.touched)
+				.some(isTouched => isTouched)
+			break
+
+		default:
+			console.warn(`Unrecognized action dispatched: ${type}`, payload)
+			return state
+	}
+
+	return newState
+}
+
+
+
+
+
+const FormContext = createContext({
+	...INITIAL_STATE,
+	updateValidity: () => {},
+	updateValue: () => {},
+})
+
+const Form = props => {
+	const {
+		children,
+		initialValues,
+		onSubmit,
+	} = props
+	const [state, dispatch] = useReducer(reducer, {
+		...INITIAL_STATE,
+		errors: Object
+			.keys(initialValues)
+			.reduce((accumulator, key) => {
+				accumulator[key] = []
+				return accumulator
+			}, {}),
+		initialValues,
+		touched: Object
+			.keys(initialValues)
+			.reduce((accumulator, key) => {
+				accumulator[key] = false
+				return accumulator
+			}, {}),
+		validity: Object
+			.keys(initialValues)
+			.reduce((accumulator, key) => {
+				accumulator[key] = false
+				return accumulator
+			}, {}),
+		values: { ...initialValues },
+	})
+
+	const handleSubmit = useCallback(event => {
+		event.preventDefault()
+		onSubmit(state)
+	}, [
+		onSubmit,
```

Comment:
This is a pretty clever use of memoization that took me a minute to understand because I've never used parameters in the dependency chain. To make sure I understand, this makes it so our onsubmit callback returns a memoized value if the state has not changed since the last time the onsubmit function has been called. But if state or onsubmit has changed, it will re-compute correct?

Reply:
Yup, you nailed it! That's also why we always return a cloned and updated object in the reducer, rather than just updating the original state and returning it. If we create a new object, then it looks like the entire state has changed. If we were to only update values within the object, then the `handleSubmit` function would never update.